src/clipping/weileratherton.jl

```diff
@@ -0,0 +1,203 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    WeilerAtherton()
+
+The Weiler-Atherton algorithm for clipping polygons.
+
+## References
+
+* Weiler, Kevin and Atherton, Peter. 1977. [Hidden surface removal using polygon area sorting]
+  (https://dl.acm.org/doi/10.1145/563858.563896)
+"""
+struct WeilerAtherton <: ClippingMethod end
+
+function clip(poly::Polygon{Dim,T}, other::Geometry{Dim,T}, method::WeilerAtherton) where {Dim, T}
+  if hasholes(poly)
+    outers = clip(rings(poly)[1], boundary(other), method)
+    inners = vcat([clip(r, boundary(other), method) for r in rings(poly)[2:end]]...)
+    if isempty(outers)
+      return nothing
+    end
+
+    g = PolyArea{Dim,T}[]
+    # build a PolyArea to each outer ring generated
+    for o in outers
+      ringsₒ = [o]
+      # find inner rings inside the outer ring o 
+      for i in inners
+        if any([sideof(v, o) == IN for v in vertices(i)])
+          push!(ringsₒ, i)
+        end
+      end
+      push!(g, PolyArea(ringsₒ))
+    end
+    Multi(g)
+  else
+    outers = clip(rings(poly)[1], boundary(other), method)
+    if isempty(outers)
+      nothing
+    else
+      Multi(PolyArea.(outers))
+    end
+  end
+end
+
+function clip(ring::Ring{Dim,T}, other::Ring{Dim,T}, ::WeilerAtherton) where {Dim,T}
+  vᵣ = orientation(ring) == CCW ? vertices(ring) : reverse(vertices(ring))
+  vₒ = orientation(other) == CCW ? vertices(other) : reverse(vertices(other))
+
+  I, pᵣ, tᵣ, pₒ, tₒ = _pairwiseintersections(vᵣ, vₒ)
```

Comment:
A function that returns more than 4 variables is an indication of sub-optimal design. Can you use the a different data structure to write this algorithm? Simple lists should be enough? It is hard to grasp the main idea of the algorithm when it has so much indirection into hash tables.

The idea should be as simple as:

1. Find all pairwise intersections and save two circular lists of points.
2. Associate to each list a vector of Bool telling if the point is intersection
3. Close the loops with the algorithm for all rings of the polygon
4. Query inner rings to find which outer loop they belong too

If these for steps are not straightforward to read in the code, it is a sign that we can refactor to make it cleaner and easier to read and maintain.